pkg/scheduler/scheduler.go

```diff
@@ -324,23 +324,51 @@ type partialAssignment struct {
 	preemptionTargets []*workload.Info
 }
 
+// limitUsage makes sure that cluster queue usage does not exceed nominal+borrowing quota
+func limitUsage(e flavorassigner.Assignment, cq *cache.ClusterQueue) flavorassigner.Assignment {
+	for flavor, resourceUsage := range e.Usage {
+		cqQuota := make(map[corev1.ResourceName]*cache.ResourceQuota)
+		for _, rg := range cq.ResourceGroups {
+			for _, cqFlavor := range rg.Flavors {
+				if cqFlavor.Name == flavor {
+					cqQuota = cqFlavor.Resources
+				}
+			}
+		}
+		for resource, usage := range resourceUsage {
+			if cqQuota[resource].BorrowingLimit != nil {
+				usage = min(usage, cqQuota[resource].Nominal+*cqQuota[resource].BorrowingLimit-cq.Usage[flavor][resource])
+			}
+			resourceUsage[resource] = usage
+		}
+		e.Usage[flavor] = resourceUsage
+	}
+	return e
+}
+
 func (s *Scheduler) getAssignments(log logr.Logger, wl *workload.Info, snap *cache.Snapshot) (flavorassigner.Assignment, []*workload.Info) {
 	cq := snap.ClusterQueues[wl.ClusterQueue]
 	fullAssignment := flavorassigner.AssignFlavors(log, wl, snap.ResourceFlavors, cq, nil)
-	var fullAssignmentTargets []*workload.Info
+	var preemtionTargets []*workload.Info
 
 	arm := fullAssignment.RepresentativeMode()
 	if arm == flavorassigner.Fit {
 		return fullAssignment, nil
 	}
 
 	if arm == flavorassigner.Preempt {
-		fullAssignmentTargets = s.preemptor.GetTargets(*wl, fullAssignment, snap)
+		preemtionTargets = s.preemptor.GetTargets(*wl, fullAssignment, snap)
+	}
+
+	// if no preemtion and borrowing is going to happen, we should limit
+	// assignment usage, so it's won't book extra resources.
+	if len(preemtionTargets) == 0 && !fullAssignment.Borrowing {
+		fullAssignment = limitUsage(fullAssignment, cq)
```

Comment:
This looks like an odd place to call this function. We are saying that the workload is requesting less resources than it actually is requesting, corrupting logs and possibly affecting other calculations down the line. This logic instead should be called somewhere in `totalUsageForCommonFlavorResources`. And I think the conclusion was that we should limit it up to the `nominal`, not `nominal+borrowingLimit`.

Also this solution wouldn't be enough when partial admission is possible.

Reply:
Updated